### ISSUE TITLE
Improved model elements creation in tree 

### ISSUE THREAD

**Issue by sz332** [gaphor/gaphor#2606]
### Is your feature request related to a problem?

This is an issue where we can collect all the needs to improve model elements creation in the tree.

**Comment by sz332** [comment]
It would be nice to have a hotkey for displaying the create new element menu for elements

**Comment by sz332** [comment]
Add New Interface Block in case of SysML.

**Comment by sz332** [comment]
New profile and new stereotype create the same icon.

**Comment by shobeira** [comment]
I am not sure if it is a good idea, but it helps to create models based on standards, e.g: limit the element type can be added to the diagram based on SysML definitions.

**Comment by sz332** [comment]
> I am not sure if it is a good idea, but it helps to create models based on standards, e.g: limit the element type can be added to the diagram based on SysML definitions.

I think it works that way: only those elements can be added when you select SySML profile which are in SysML (and even there is a limited set). Is this what you need?

**Comment by shobeira** [comment]
No. Only certain type of elements are allowed in each of the SysML diagrams according to the standard.

The table attached is from the book "SysML Distilled: A Brief Guide to the Systems Modeling Language" by Lenny Delligatti.


![Allowable_model_elements_on_Diagrams](https://github.com/gaphor/gaphor/assets/43518764/f7bd875b-850b-497b-bf76-6818ca91a2a1)

**Comment by sz332** [comment]
I see, I think we need to check the specification as well and see whether they are enforced/enforcable limitations or not. One thing that came in my mind is displaying requirements on block definition diagrams showing requirements being satisfied by blocks for example, but I get your point.

**Comment by shobeira** [comment]
Based on my experience, the limitation of not being able to place element of choice into a diagram is bothering and it is not a good idea to apply the limitations as it is very useful for illustration purposes and modeling. It is problematic in model simulation though.

The decision could be made by Gaphor direction: whether it will develop for simulation or it is just an illustration software.

I prefer it to be my illustration software and give me the freedom to design diagrams the way I need to.

**Comment by sz332** [comment]
I totally agree with you, to solve this problem a validator might be more sufficient (ensuring company level rules and regulations).

How can we connect this discussion to model element creation in the model tree? Shall we create a new one feature request and continue there, or can we connect it to this feature anyhow?

**Comment by shobeira** [comment]
The user should be able to add elements without any limitation to the model tree, but may not be able to place ANY element to a specific diagram (if the Gaphor directors decide to go in the simulation features). The discussion is not related to the element creation topic in my idea as we should have improvement ideas about the element creation in the model tree.

**Comment by marek-piirikivi** [comment]
> No. Only certain type of elements are allowed in each of the SysML diagrams according to the standard.

@shobeira This Table 2.1 is about what elements can own which diagrams and is already implemented about a week or two ago by me. It's a first information on the diagram header. For example `[block] My block name [Diagram name]` means that the diagram "Diagram name" is owned by a block named "My block name". Another example on what it means is that a package can no longer hold activity diagram - a new activity is created for you if you try to do so.

But you are overall onto something - there is indeed a list somewhere about which element types can own what. For example, I know that a requirement can only be created under a package or another requirement (maybe something more, but I am not 100% sure). Similarly how there are definitely other kinds of elements that can be own only by a specific set of elements.

Would this be something that you are after?

**Comment by marek-piirikivi** [comment]
If we talk about the gaphor being only illustration software; I think a generic diagram would already serve this purpose.

Also, I think that deferred validation would work best - a tool that would highlight what you have incorrect; instead of blatantly preventing from drawing the diagrams freely. And add good defaults (which are already correct) to that - and I think you have a quite decent combination of being correct and having freedom.

**Comment by shobeira** [comment]
> > No. Only certain type of elements are allowed in each of the SysML diagrams according to the standard.
> 
> @shobeira This Table 2.1 is about what elements can own which diagrams and is already implemented about a week or two ago by me. It's a first information on the diagram header. For example `[block] My block name [Diagram name]` means that the diagram "Diagram name" is owned by a block named "My block name". Another example on what it means is that a package can no longer hold activity diagram - a new activity is created for you if you try to do so.
> 
> But you are overall onto something - there is indeed a list somewhere about which element types can own what. For example, I know that a requirement can only be created under a package or another requirement (maybe something more, but I am not 100% sure). Similarly how there are definitely other kinds of elements that can be own only by a specific set of elements.
> 
> Would this be something that you are after?

What I understand from the Table 2.1 and other reliable materials is that every diagram could have certain types of elements. Although if Gaphor direction is to be an illustration only software, the freedom of drawing is in higher priority than validating the diagrams according to standards.

**Comment by marek-piirikivi** [comment]
> What I understand from the Table 2.1 and other reliable materials

I would be cautious on any materials that are not official standard. The same table in the SysMLv1.6 specification is described as the following (screenshot from the standard)

![image](https://github.com/gaphor/gaphor/assets/45357878/3d040884-7189-4995-8971-7d4b507874e6)

It tells you only which element types are allowed diagram element types; but not which elements can be present in the diagram.

**Comment by shobeira** [comment]
@marek-piirikivi,

Please have a look at page #213 of the official standard where the above material posted. it stated:

> The following are **some of** the designated model elements associated with the different diagram kinds.

While the lines before posted screenshot stated "some of the sesignated model elements" with no enforcement, the table 2.1 from the book "SysML Distilled: A Brief Guide to the Systems Modeling Language" clearly illistrated allowable model element types for each diagram kinds.

**Comment by marek-piirikivi** [comment]
@shobeira I feel as if we are talking about slightly different things.

I am talking about the element which the diagram can represent - which is defined by the table that you also showed. This is already implemented (with exception to "model", "modelView" elements, and maybe few others).

And I feel that you are talking about the contents of the diagram, not about the single element which internals the diagram explains. Basically telling that "action should only be creatable under activity in the model browser" etc.

Could you confirm/fix my understanding?
